src/core/device.c

```diff
@@ -190,18 +190,19 @@ static int device_coldplug(Unit *u) {
          * - On switch-root, the udev database may be cleared, except for devices with sticky bit, i.e.
          *   OPTIONS="db_persist". Hence, almost no devices are enumerated in the step 2. However, in general,
          *   we have several serialized devices. So, DEVICE_FOUND_UDEV bit in the deserialized_found must be
-         *   ignored, as udev rules in initramfs and the main system are often different. If the deserialized
-         *   state is DEVICE_PLUGGED, we need to downgrade it to DEVICE_TENTATIVE (or DEVICE_DEAD if nobody
-         *   sees the device). Unlike the other starting mode, MANAGER_IS_SWITCHING_ROOT() is true when
-         *   device_coldplug() and device_catchup() are called. Hence, let's conditionalize the operations by
-         *   using the flag. After switch-root, systemd-udevd will (re-)process all devices, and the
-         *   Device.found and Device.state will be adjusted.
+         *   ignored unless the device is enumerated by device_enumerate(), as udev rules in initramfs and
+         *   the main system are often different. If the deserialized state is DEVICE_PLUGGED, we need to
+         *   downgrade it to DEVICE_TENTATIVE (or DEVICE_DEAD if nobody sees the device). Unlike the other
+         *   starting mode, MANAGER_IS_SWITCHING_ROOT() is true when device_coldplug() and device_catchup()
+         *   are called. Hence, let's conditionalize the operations by using the flag. After switch-root,
+         *   systemd-udevd will (re-)process all devices, and the Device.found and Device.state will be
+         *   adjusted.
          *
          * - On reload or reexecute, we can trust enumerated_found, deserialized_found, and deserialized_state.
          *   Of course, deserialized parameters may be outdated, but the unit state can be adjusted later by
          *   device_catchup() or uevents. */
 
-        if (MANAGER_IS_SWITCHING_ROOT(m)) {
+        if (MANAGER_IS_SWITCHING_ROOT(m) && !FLAGS_SET(d->enumerated_found, DEVICE_FOUND_UDEV)) {
```

Comment:
if you are able to reproduce the issue, can you confirm whether this fixes it for you?

Reply:
Thank you.

Reply:
I'm not seeing these "dead->plugged" transitions anymore right after switching root for devices with the "db_persist" option set.

Reply:
That's one possible intepretation of the text in **udev(7)**. But you're just inventing it :-)

`db_persist` dates back to 2011, before the systemd/udev merge. The man page paragraph was added much later. I don't think the relationship of `db_persist` and systemd's `state` attribute has ever been clearly defined. 

`db_persist` was created in the past for LVM, because LVM uses "change" events for device activation rather than "add" events, and needs special treatment for coldplug "add" events for devices that are already active (see `10-dm.rules`). LVM doesn't care about systemd's `state`, so as far as LVM is concerned, it doesn't matter whether we translate `db_persist` to `DEVICE_READY` or `DEVICE_TENTATIVE`.

I still think `DEVICE_TENTATIVE` is safer. It doesn't do harm, because the actual state will be resolved sooner or later either way, and it avoids premature activation.

Reply:
> I think this is wrong. We must reset the state to `DEVICE_TENTATIVE` to prevent premature device activation in the switching root case. The argument is the same as in #23208. Even with `db_persist`, the udev rules may have changed while switching root, and a device which was "plugged" in the initramfs may not be any more in the mounted root (SYSTEMD_READY="0" may be set on it). IMO `DEVICE_READY` should never be set after switching root until an actual "add" uevent for the device has been processed.

@mwilck The host PID1 has used and honored the persistent database generated in initrd at least more than two years. Why do you want to change the behavior? Of course, we have not documented the behavior, but IIRC, there is no issue related to that.
As reported in #23429, after PR #23218, the device with persistent database has unnecessary state transition. It may be harmless in most cases, but such state transition did not occur previously. If we can avoid that, why not to do that?

> Well if db_persist is used it's clearly an indication that the device must remain plugged, see the doc. If SYSTEMD_READY=0 is used in the host then it seems actually expected that the device goes through the `plugged->dead` transition when the new set of rules from the host is processed by udev.

@fbuihuu I agree with you. Even if the behavior is not documented, we have done so for a while. There is no reason to change it.

Reply:
> @mwilck The host PID1 has used and honored the persistent database generated in initrd at least more than two years. Why do you want to change the behavior?

@yuwata, you're right that as a matter of fact, systemd kept `DEVICE_READY` state over switching root. It did this for **all** devices though, which was wrong, and was fixed in #23218. #23218 did indeed "change the behavior". I hope we agree it did so for the better. 

The open question is how to deal with `db_persist`. We need to realize that we have two conflicting semantic rules:

 1. udev rules are allowed to be different between initrd and root fs, and thus need to be re-parsed and re-executed after switching root before switching device states and forwarding them to dependent units.
 2. `db_persist` devices keep state over switching root.

@fbuihuu and you think that rule 2) defines an exception to rule 1). I disagree. 1) is a general principle. **Nothing** prevents udev rules from changing between initrd and root FS. Thus if we run `device_set_state(DEVICE_READY)` on these devices, we'd open up new ways for the race described in #23208 to occur, which may cause boot failure.

`db_persist` is only used by dm and md. After switching root, the  udev rules for these subsystems need to distinguish between "activation" events and other events, which requires certain udev properties to survive the cleanup (see e.g. [this dracut commit](https://github.com/dracutdevs/dracut/commit/b874d4b5be4c3b81cf394726c354ef7131b5699d)). This distinction is made in udev rules and relies 100% on properties in the udev db. It has nothing to do with systemd's device activation. Thus for the current consumers of `db_persist`, it doesn't matter if we set the state to "tentative" rather than "plugged" here. 

It does matter for the further processing of dependent devices, though. The processing of dm and md devices during boot relies heavily on (udev) coldplug. It's not in the interest of these subsystems to have dependent units activated before the (udev) coldplug events have been processed.  Strictly speaking, after switching root, we should never call `device_set_state(DEVICE_READY)`, except from `device_dispatch_io()`.

**However**, I realize that this hunk is not the point. With this hunk, we'll set `DEVICE_READY` in `device_coldplug()` for `db_persist` devices, whereas without it, we set it in `device_catchup()`. If my argument above is correct, both calls would be premature. @fbuihuu, with this observation I don't understand [how this patch fixes your issue](https://github.com/systemd/systemd/pull/23437#discussion_r877134440). The extra "dead→plugged" transition should still observed, just a little bit earlier.